File: android/app/src/test/java/com/tappyai/app/share/ShareArtifactTest.kt
package com.tappyai.app.share

import com.tappyai.app.chat.PlanDay
import com.tappyai.app.chat.PlanItem
import com.tappyai.app.chat.TappyPlan
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/**
 * The share artifact is a WHITELIST. These tests pin what may leave the app, what may not,
 * and that the brochure is deterministic and bounded for the web Inbox (chat_messages.body ≤ 4000).
 */
class ShareArtifactTest {

    private val line = File("src/test/resources/share/places_annotation_line.txt").readText().trim()
    private val view = PlacesLiveViewParser.fromStreamLine(line)!!

    @Test
    fun `brochure has the TappyAI header, every place, its facts, and the brand footer`() {
        val a = ShareArtifactBuilder.buildPlacesArtifact(view, "Quán bún bò ngon ở TP.HCM", "vi")
        assertEquals(ShareArtifact.Kind.PLACES, a.kind)
        assertEquals("TappyAI gợi ý: Quán bún bò ngon ở TP.HCM", a.subject)
        assertTrue(a.text.startsWith("TappyAI gợi ý: Quán bún bò ngon ở TP.HCM\n\n1. "))
        assertTrue(a.text.endsWith("Gợi ý bởi TappyAI · www.tappyai.com"))
        assertEquals("https://www.tappyai.com", a.url)
        assertEquals(8, a.places.size)
        for ((i, p) in view.items.withIndex()) {
            assertTrue(a.text.contains("${i + 1}. ${p.name}"))
            p.phone?.let { assertTrue(a.text.contains("☎ $it")) }
            p.openingHours?.let { assertTrue(a.text.contains("🕐 $it")) }
            p.address?.let { assertTrue(a.text.contains("📍 $it")) }
            val maps = p.actions.first { it.kind == "maps" && it.urlKind == "direct" }
            assertTrue(a.text.contains("Bản đồ: ${maps.url}"))
        }
        assertTrue(a.text.contains("★ 4.9 (5.946 đánh giá)"))
    }

    @Test
    fun `nothing internal can enter the text`() {
        val a = ShareArtifactBuilder.buildPlacesArtifact(view, "x", "vi")
        for (forbidden in listOf("place:osm:", "shortlistPosition", "\"rank\"", "provenance", "evidence_type", "source_type",
            "matchVerdict", "distanceKm", "priceSignal", "_tappy_", "tappy.places.v1")) {
            assertFalse("leaked $forbidden", a.text.contains(forbidden))
        }
        // Search links (ShopeeFood/GrabFood/Google searches) are not the venue's own; they stay out.
        assertFalse(a.text.contains("shopeefood.vn/search"))
        assertFalse(a.text.contains("google.com/search"))
        for (p in a.places) for (l in p.links) assertTrue(l.url.startsWith("https://"))
    }

    @Test
    fun `link labels name the platform only when it adds information`() {
        val block = ShareArtifactBuilder.placeBlock(
            SharedPlace(name = "x", links = listOf(
                SharedLink("website", "https://a.example/", "Official Website"),
                SharedLink("maps", "https://maps.google.com/?cid=1", "Google Maps"),
                SharedLink("review", "https://www.tiktok.com/@a/video/1", "TikTok"),
                SharedLink("order", "https://shopeefood.vn/x", "ShopeeFood"),
            )), 0, "vi",
        )
        assertTrue(block.contains("   Website: https://a.example/"))
        assertTrue(block.contains("   Bản đồ: https://maps.google.com/?cid=1"))
        assertTrue(block.contains("   Review (TikTok): https://www.tiktok.com/@a/video/1"))
        assertTrue(block.contains("   Đặt món (ShopeeFood): https://shopeefood.vn/x"))
        assertFalse(block.contains("Official Website"))
    }

    @Test
    fun `same input gives the same text`() {
        val a = ShareArtifactBuilder.buildPlacesArtifact(view, "x", "vi").text
        val b = ShareArtifactBuilder.buildPlacesArtifact(view, "x", "vi").text
        assertEquals(a, b)
    }

    @Test
    fun `inbox body fits 4000 without cutting a URL and says how many were dropped`() {
        // Inflate to force compaction: replicate the real list until it is far over the bound.
        val big = PlacesLiveView(view.domain, List(6) { view.items }.flatten())
        val a = ShareArtifactBuilder.buildPlacesArtifact(big, "Quán bún bò ngon ở TP.HCM", "vi")
        assertTrue(a.text.length > ShareArtifactBuilder.INBOX_MAX_BODY)
        val body = ShareArtifactBuilder.inboxBody(a, "vi")
        assertTrue(body.length <= ShareArtifactBuilder.INBOX_MAX_BODY)
        assertTrue(body.endsWith("Gợi ý bởi TappyAI · www.tappyai.com"))
        assertTrue(body.startsWith("TappyAI gợi ý: Quán bún bò ngon ở TP.HCM"))
        assertTrue(Regex("và \\d+ địa điểm khác").containsMatchIn(body))
        // Every URL in the compact body is a complete URL that exists in the full text.
        val urls = Regex("https://\\S+").findAll(body).map { it.value }.toList()
        assertTrue(urls.isNotEmpty())
        for (u in urls) assertTrue("cut url $u", a.text.contains(u + "\n") || a.text.endsWith(u))
    }

    @Test
    fun `short brochure is passed through unchanged`() {
        val a = ShareArtifactBuilder.buildPlacesArtifact(view, "x", "vi")
        if (a.text.length <= ShareArtifactBuilder.INBOX_MAX_BODY) {
            assertEquals(a.text, ShareArtifactBuilder.inboxBody(a, "vi"))
        }
    }

    @Test
    fun `plan brochure is built from the plan structure, share_text only as a short intro`() {
        val plan = TappyPlan(
            title = "1 ngày Đà Nẵng", people = 2, budgetTotal = "1.500.000₫",
            days = listOf(PlanDay(label = "Ngày 1", items = listOf(
                PlanItem(time = "08:00", emoji = "☕", name = "Cà phê Cộng", description = "Bắt đầu nhẹ", price = "60.000₫",
                    address = "96 Bạch Đằng", mapsLink = "https://maps.google.com/?q=Cong", bookingLink = "javascript:alert(1)"),
            ))),
            shareText = "Kế hoạch 1 ngày Đà Nẵng cho 2 người — xem tại https://evil.example",
        )
        val a = ShareArtifactBuilder.buildPlanArtifact(plan, "vi")
        assertEquals(ShareArtifact.Kind.PLAN, a.kind)
        assertTrue(a.text.startsWith("Kế hoạch từ TappyAI: 1 ngày Đà Nẵng\n"))
        assertFalse(a.text.contains("evil.example"))
        assertTrue(a.text.contains("2 người · Ngân sách: 1.500.000₫"))
        assertTrue(a.text.contains("Ngày 1\n  08:00 ☕ Cà phê Cộng\n     Bắt đầu nhẹ\n     60.000₫ · 📍 96 Bạch Đằng\n     Bản đồ: https://maps.google.com/?q=Cong"))
        assertFalse(a.text.contains("javascript:"))
        assertTrue(a.text.endsWith("Gợi ý bởi TappyAI · www.tappyai.com"))
    }

    @Test
    fun `prose fallback still carries the brand`() {
        val a = ShareArtifactBuilder.buildProseArtifact("hỏi gì đó", "Câu trả lời.")
        assertTrue(a.text.startsWith("TappyAI\n\n"))
        assertTrue(a.text.endsWith("— TappyAI · tappyai.com"))
        assertTrue(a.text.contains("Câu trả lời."))
    }

    @Test
    fun `prose keeps safe links as label colon url and drops images and unsafe links`() {
        val a = ShareArtifactBuilder.buildProseArtifact(
            "máy bay",
            "**Gợi ý**: đặt qua [Traveloka](https://www.traveloka.com/vi-vn) · [Evil](javascript:alert(1)) ![ảnh](https://img.example/a.jpg)\n\nXem thêm https://www.vietnamairlines.com/ nhé http://insecure.example/x",
        )
        assertTrue(a.text.startsWith("TappyAI\n\nGợi ý: đặt qua Traveloka: https://www.traveloka.com/vi-vn · Evil"))
        assertFalse(a.text.contains("javascript:"))
        assertFalse(a.text.contains("img.example"))
        assertFalse(a.text.contains("**"))
        assertTrue(a.text.contains("https://www.vietnamairlines.com/"))
        assertFalse(a.text.contains("insecure.example"))
        assertTrue(a.text.endsWith("— TappyAI · tappyai.com"))
    }

    @Test
    fun `english labels when the locale is english`() {
        val a = ShareArtifactBuilder.buildPlacesArtifact(view, "beef", "en")
        assertTrue(a.text.startsWith("TappyAI recommends: beef"))
        assertTrue(a.text.contains("Maps: https://"))
        assertTrue(a.text.endsWith("Recommended by TappyAI · www.tappyai.com"))
        assertTrue(a.text.contains("(5,946 reviews)"))
    }
}

File: src/lib/share/shareArtifact.test.ts
// The share artifact — the ONE thing that leaves TappyAI when a user shares a
// recommendation or a plan.
//
// It is a WHITELIST. `SharedPlace` names every field allowed out; `pickPlace`
// copies exactly those. The tests below feed the builder a REAL `8:` annotation
// captured from `/api/chat` (not a hand-made fixture that is cleaner than
// production) and prove: every user-facing fact is present, nothing internal
// can appear, the text is deterministic, the brand footer is there, and the
// Inbox body fits `chat_messages.body ≤ 4000` without ever cutting a URL.

import { describe, it, expect } from 'vitest'
import type { LivePlace, PlacesLiveView } from '@/lib/recommendation/liveView'
import type { TappyPlan } from '@/components/TripPlanCard'
import fixture from './__fixtures__/placesLiveView.food.json'
import {
  INBOX_MAX_BODY,
  artifactUrls,
  buildPlacesArtifact,
  buildPlanArtifact,
  buildProseArtifact,
  compactBrochure,
  inboxBody,
  pickPlace,
  placeBlock,
  type SharedPlace,
} from './shareArtifact'
import { isShareableUrl } from './shareTargets'

const view = fixture as unknown as PlacesLiveView
const env = { NEXT_PUBLIC_SITE_URL: 'https://www.tappyai.com' } as unknown as NodeJS.ProcessEnv
const TITLE = 'Quán bún bò ngon ở TP.HCM'

// ------------------------------------------------------------ the fixture
describe('fixture is a real production frame', () => {
  it('is the tappy.places.v1 annotation with eight places', () => {
    expect(view.kind).toBe('tappy.places.v1')
    expect(view.items).toHaveLength(8)
    // The fields the prose stream does NOT carry — the reason mobile parses `8:`.
    expect(view.items[0].phone).toBeTruthy()
    expect(view.items[0].openingHours).toBeTruthy()
    expect(view.items[0].actions.some((a) => a.kind === 'maps' && a.urlKind === 'direct')).toBe(true)
    // And the internal fields the whitelist must keep out.
    expect(view.items[0].id).toMatch(/^place:/)
    expect(typeof view.items[0].rank).toBe('number')
  })
})

// --------------------------------------------------------------- whitelist
describe('pickPlace — whitelist', () => {
  it('copies only the named fields', () => {
    const p = pickPlace(view.items[0])
    expect(Object.keys(p).sort()).toEqual(
      ['address', 'category', 'image', 'links', 'name', 'openingHours', 'phone', 'rating', 'ratingCount'].sort()
    )
  })

  // The privacy field and the internal fields have NO slot to land in — even
  // when they are present on the input.
  it('cannot carry distanceKm, rank, id, matchVerdict, priceSignal or provenance', () => {
    const poisoned = {
      ...view.items[0],
      distanceKm: 0.9,
      matchVerdict: 'match',
      priceSignal: '50k',
      shortlistPosition: 1,
      provenance: { source_type: 'serper_places', evidence_type: 'listing' },
      _tappy_debug: 'x',
    } as unknown as LivePlace
    const p = pickPlace(poisoned) as unknown as Record<string, unknown>
    for (const k of ['distanceKm', 'rank', 'id', 'matchVerdict', 'priceSignal', 'shortlistPosition', 'provenance', '_tappy_debug', 'domain', 'kind']) {
      expect(p).not.toHaveProperty(k)
    }
    const text = placeBlock(p as unknown as SharedPlace, 0, 'vi')
    expect(text).not.toMatch(/0\.9|match|50k|serper|listing|_tappy_/)
  })

  it('keeps only direct https links of shareable kinds, deduplicated', () => {
    const p = pickPlace(view.items[0])
    for (const l of p.links) {
      expect(l.url).toMatch(/^https:\/\//)
      expect(['maps', 'website', 'review', 'order', 'booking', 'ticket', 'reservation']).toContain(l.kind)
    }
    expect(new Set(p.links.map((l) => l.url)).size).toBe(p.links.length)
    // Search links are the platform's search, not the venue — they stay out.
    // Checked against the ACTUAL search URLs the frame carries, not a guessed pattern.
    const searchUrls = view.items[0].actions.filter((a) => a.urlKind === 'search').map((a) => a.url)
    expect(searchUrls.length).toBeGreaterThan(0)
    for (const u of searchUrls) expect(p.links.some((l) => l.url === u)).toBe(false)
    expect(p.links.map((l) => l.url).sort()).toEqual(
      [...new Set(view.items[0].actions.filter((a) => a.urlKind === 'direct' && a.url.startsWith('https://')).map((a) => a.url))].sort()
    )
    // `call` is the phone field, not a link.
    expect(p.links.some((l) => (l.kind as string) === 'call')).toBe(false)
  })

  it('drops a javascript: or http: link even if it arrived as direct', () => {
    const bad = {
      ...view.items[0],
      actions: [
        { kind: 'maps', urlKind: 'direct', url: 'javascript:alert(1)', labelKey: 'x' },
        { kind: 'website', urlKind: 'direct', url: 'http://insecure.example', labelKey: 'x' },
        { kind: 'maps', urlKind: 'direct', url: 'https://maps.google.com/?cid=1', labelKey: 'x' },
      ],
    } as unknown as LivePlace
    expect(pickPlace(bad).links.map((l) => l.url)).toEqual(['https://maps.google.com/?cid=1'])
  })
})

describe('placeBlock — link labels', () => {
  it('names the platform when it adds information, and not when it only restates the kind', () => {
    const block = placeBlock({
      name: 'x',
      links: [
        { kind: 'website', url: 'https://a.example/', platform: 'Official Website' },
        { kind: 'maps', url: 'https://maps.google.com/?cid=1', platform: 'Google Maps' },
        { kind: 'review', url: 'https://www.tiktok.com/@a/video/1', platform: 'TikTok' },
        { kind: 'order', url: 'https://shopeefood.vn/x', platform: 'ShopeeFood' },
      ],
    }, 0, 'vi')
    expect(block).toContain('   Website: https://a.example/')
    expect(block).toContain('   Bản đồ: https://maps.google.com/?cid=1')
    expect(block).toContain('   Review (TikTok): https://www.tiktok.com/@a/video/1')
    expect(block).toContain('   Đặt món (ShopeeFood): https://shopeefood.vn/x')
    expect(block).not.toContain('Official Website')
  })
})

// ---------------------------------------------------------------- brochure
describe('buildPlacesArtifact — the brochure', () => {
  const a = buildPlacesArtifact(view, TITLE, 'vi', env)

  it('opens with the TappyAI header and closes with the brand footer', () => {
    expect(a.subject).toBe(`TappyAI gợi ý: ${TITLE}`)
    expect(a.text.startsWith(`TappyAI gợi ý: ${TITLE}\n\n1. `)).toBe(true)
    expect(a.text.endsWith('Gợi ý bởi TappyAI · www.tappyai.com')).toBe(true)
    expect(a.url).toBe('https://www.tappyai.com')
    expect(isShareableUrl(a.url, env)).toBe(true)
  })

  it('carries every place with its facts — no information loss vs the card', () => {
    expect(a.places).toHaveLength(8)
    view.items.forEach((p, i) => {
      expect(a.text).toContain(`${i + 1}. ${p.name}`)
      if (p.address) expect(a.text).toContain(`📍 ${p.address}`)
      if (p.phone) expect(a.text).toContain(`☎ ${p.phone}`)
      if (p.openingHours) expect(a.text).toContain(`🕐 ${p.openingHours}`)
      const maps = p.actions.find((x) => x.kind === 'maps' && x.urlKind === 'direct')
      if (maps) expect(a.text).toContain(`Bản đồ: ${maps.url}`)
    })
    expect(a.text).toContain('★ 4.9 (5.946 đánh giá)')
  })

  it('leaks nothing internal', () => {
    for (const forbidden of ['place:osm:', 'shortlistPosition', 'rank', 'provenance', 'evidence_type', 'source_type',
      'matchVerdict', 'distanceKm', 'priceSignal', '_tappy_', 'tappy.places.v1', 'openingHoursWeek']) {
      expect(a.text).not.toContain(forbidden)
    }
    expect(JSON.stringify(a.places)).not.toMatch(/distanceKm|provenance|matchVerdict|priceSignal|"rank"|"id"/)
  })

  it('is deterministic', () => {
    expect(buildPlacesArtifact(view, TITLE, 'vi', env).text).toBe(a.text)
  })

  it('every URL in the text is https and never a private TappyAI route', () => {
    const urls = artifactUrls(a)
    expect(urls.length).toBeGreaterThan(0)
    for (const u of urls) {
      expect(u).toMatch(/^https:\/\//)
      expect(u).not.toMatch(/tappyai\.com\/(api|chat|admin|auth|login)/)
    }
  })

  it('speaks English when asked', () => {
    const en = buildPlacesArtifact(view, 'beef', 'en', env)
    expect(en.text.startsWith('TappyAI recommends: beef')).toBe(true)
    expect(en.text).toContain('Maps: https://')
    expect(en.text).toContain('(5,946 reviews)')
    expect(en.text.endsWith('Recommended by TappyAI · www.tappyai.com')).toBe(true)
  })
})

// ------------------------------------------------------------- compaction
describe('inboxBody — Tappy Inbox ≤ 4000, URLs intact', () => {
  const big: PlacesLiveView = { ...view, items: Array.from({ length: 6 }, () => view.items).flat() }
  const a = buildPlacesArtifact(big, TITLE, 'vi', env)

  it('the inflated brochure really overflows', () => {
    expect(a.text.length).toBeGreaterThan(INBOX_MAX_BODY)
  })

  it('fits, keeps header + footer, says how many were dropped', () => {
    const body = inboxBody(a, 'vi')
    expect(body.length).toBeLessThanOrEqual(INBOX_MAX_BODY)
    expect(body.startsWith(`TappyAI gợi ý: ${TITLE}`)).toBe(true)
    expect(body.endsWith('Gợi ý bởi TappyAI · www.tappyai.com')).toBe(true)
    expect(body).toMatch(/và \d+ địa điểm khác/)
  })

  it('never cuts a URL — every URL in the body is a whole URL from the full text', () => {
    const body = inboxBody(a, 'vi')
    const urls = body.match(/https:\/\/\S+/g) ?? []
    expect(urls.length).toBeGreaterThan(0)
    for (const u of urls) expect(a.text.includes(`${u}\n`) || a.text.endsWith(u)).toBe(true)
  })

  it('passes a short brochure through unchanged', () => {
    const small = buildPlacesArtifact({ ...view, items: view.items.slice(0, 2) }, TITLE, 'vi', env)
    expect(inboxBody(small, 'vi')).toBe(small.text)
  })

  it('holds the bound even for a single enormous place', () => {
    const monster = { ...view.items[0], address: 'x'.repeat(5000) } as LivePlace
    const out = compactBrochure(TITLE, [pickPlace(monster)], 'vi', 'https://www.tappyai.com')
    expect(out.length).toBeLessThanOrEqual(INBOX_MAX_BODY)
    expect(out).toContain(`1. ${monster.name}`)
  })
})

// ------------------------------------------------------------------- prose
describe('buildProseArtifact — a turn with no card and no plan', () => {
  it('keeps safe links as "label: url", drops images and unsafe links, and carries the brand', () => {
    const a = buildProseArtifact(
      'máy bay đi Đà Nẵng',
      '**Gợi ý**: đặt qua [Traveloka](https://www.traveloka.com/vi-vn) · [Evil](javascript:alert(1)) ![ảnh](https://img.example/a.jpg)\n\nXem thêm https://www.vietnamairlines.com/ nhé http://insecure.example/x',
      env,
    )
    expect(a.subject).toBe('TappyAI: máy bay đi Đà Nẵng')
    expect(a.text.startsWith('TappyAI\n\nGợi ý: đặt qua Traveloka: https://www.traveloka.com/vi-vn · Evil')).toBe(true)
    expect(a.text).not.toContain('javascript:')
    expect(a.text).not.toContain('img.example')
    expect(a.text).not.toContain('**')
    expect(a.text).toContain('https://www.vietnamairlines.com/')
    expect(a.text).not.toContain('insecure.example')
    expect(a.text.endsWith('— TappyAI · tappyai.com')).toBe(true)
    expect(a.places).toEqual([])
  })
})

// -------------------------------------------------------------------- plan
describe('buildPlanArtifact — deterministic from the plan structure', () => {
  const plan: TappyPlan = {
    type: 'trip',
    title: '1 ngày Đà Nẵng',
    people: 2,
    budget_total: '1.500.000₫',
    share_text: 'Kế hoạch 1 ngày Đà Nẵng cho 2 người — xem tại https://evil.example',
    days: [{
      label: 'Ngày 1',
      items: [{
        time: '08:00', emoji: '☕', category: 'food', name: 'Cà phê Cộng', description: 'Bắt đầu nhẹ',
        price: '60.000₫', address: '96 Bạch Đằng', maps_link: 'https://maps.google.com/?q=Cong',
        booking_link: 'javascript:alert(1)', place_id: 'ChIJ-secret',
      }],
    }],
  }

  it('builds the brochure from days/items; share_text with a URL is not copied', () => {
    const a = buildPlanArtifact(plan, 'vi', env)
    expect(a.kind).toBe('plan')
    expect(a.text.startsWith('Kế hoạch từ TappyAI: 1 ngày Đà Nẵng\n')).toBe(true)
    expect(a.text).not.toContain('evil.example')
    expect(a.text).toContain('2 người · Ngân sách: 1.500.000₫')
    expect(a.text).toContain('Ngày 1\n  08:00 ☕ Cà phê Cộng\n     Bắt đầu nhẹ\n     60.000₫ · 📍 96 Bạch Đằng\n     Bản đồ: https://maps.google.com/?q=Cong')
    expect(a.text).not.toContain('javascript:')
    expect(a.text).not.toContain('ChIJ-secret')
    expect(a.text.endsWith('Gợi ý bởi TappyAI · www.tappyai.com')).toBe(true)
  })

  it('lets a short, URL-free share_text be the intro line only', () => {
    const a = buildPlanArtifact({ ...plan, share_text: 'Một ngày thật chill ở Đà Nẵng!' }, 'vi', env)
    expect(a.text.split('\n')[1]).toBe('Một ngày thật chill ở Đà Nẵng!')
  })
})
